File: compiler/bench/narcissistic.c
// expected: 20
// Narcissistic numbers benchmark - count narcissistic 1-1000000, run 5 times

#include <stdio.h>

int count_digits(int n) {
    int count = 0;
    while (n) { n /= 10; count++; }
    return count;
}

long ipow(int base, int exp) {
    if (exp == 0) return 1;
    long result = 1;
    for (int i = 0; i < exp; i++) {
        result *= base;
    }
    return result;
}

int narcissistic(int n) {
    int digits = count_digits(n);
    long sum = 0;
    int tmp = n;
    while (tmp) {
        int d = tmp % 10;
        sum += ipow(d, digits);
        tmp /= 10;
    }
    return sum == n;
}

int count_narcissistic(int limit) {
    int count = 0;
    for (int i = 1; i <= limit; i++) {
        if (narcissistic(i)) count++;
    }
    return count;
}

int main(void) {
    int result = 0;
    for (int i = 0; i < 5; i++) {
        result = count_narcissistic(1000000);
    }
    printf("%d\n", result);
    return 0;
}

File: compiler/bench/narcissistic.fs
\ expected: 20
\ Narcissistic numbers benchmark - count narcissistic 1-1000000, run 5 times

: count-digits ( n -- count )
  0 swap
  begin dup while 10 / swap 1+ swap repeat drop ;

: ipow ( base exp -- result )
  dup 0= if 2drop 1 exit then
  1 swap 0 do over * loop nip ;

: narcissistic? ( n -- flag )
  dup count-digits
  0 rot
  begin dup while
    10 /mod swap
    3 pick ipow
    rot + swap
  repeat drop
  swap drop = ;

: count-narcissistic ( limit -- count )
  0 swap 1+ 1 do
    i narcissistic? if 1+ then
  loop ;

: main ( -- )
  0
  5 0 do
    drop 1000000 count-narcissistic
  loop
  . cr ;
